accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,245 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
 It is a warning to use the null check operator (`!`) on an expression of type
 `T` if `T` is **strictly non-nullable** .
 
+### Local variables and definite (un)assignment.
+
+As part of the null safety release, errors for local variables are specified to
+take into account **definite assignment** and **definite unassignment** (see
+the section on Definite Assignment below).
+
+In all cases in this section, errors that are described as occurring on reads of
+a variable are intended to apply to all form of reads, including indirectly as
+part of composite assignment operators, as well as via pre and post-fix
+operators.  Similarly, errors that are described as occurring on writes of a
+variable are intended to apply to all form of writes.
+
+It is a compile time error to assign a value to a local variable marked `final`
+when the variable is **definitely assigned**.  This includes variables that are
+also marked `late`.  A variable is always considered definitely assigned if it
+has an explicit initializer, or an implicit initializer as part of a larger
+construct (e.g. the loop variable in a `for in` construct).
+
+It is not a compile time error to write to a `final` local variable if that
+variable is declared `late`, does not have an initializer, and is not
+**definitely assigned**.
+
+It is not a compile time error to write to a `final` local variable if that
+variable is **definitely unassigned** (and hence does not have an initializer).
+
```

Comment:
It is already specified that writing to a `final` variable is a compile time error, with certain exceptions (e.g. there is setter with the right name in scope).  This spec was intended to add to the list of exceptions, without removing the original error.  I can make this explicit though.

Reply:
Exceptions should generally be very precise about what they are exceptions to. Otherwise readers might consider them exceptions to other things as well.

I worry someone might read an "It is not a compile time error to write to a `final` local variable if that variable is **definitely unassigned**" to mean that it's also not an error if the types are wrong.

It does take some amount of disinginuity to get there, but there's always one :)

That's why I want to be specific about which compile-time error it isn't, and why I prefer not to have negative definitions at all, so exceptions are written right at the point that they are exceptions to. (So don't say "A is invalid" in one place and then say "A is valid anyway in some cases" in another place, someone is bound to miss the exception).

> It's a compile-time error to write to a local final non-late variable that is not definitely unassigned.
> It's a compile-time error to write to a local final late variable that is definitely assigned.

(Those are the only variables you write to directly outside of initializer lists, all other "variables" are getters and setters, and there it's a compile-time error if there is no setter in scope).


All in all, this is probably fine, as long as the reader is not trying to misunderstand it.

Reply:
@lrhn agreed, my intent is to be clear.  I did rewrite this based on your original comment, and I'm not sure if your last comment was intended to apply to the changed version or the original.  Would you mind taking a look at the current version and letting me know where you still see things that might be unclear?  I will do the same, but I may be too close to the proposal to do so effectively.

Reply:
The new wording is clearer and seems to cover everything.

Reply:
Great, thanks!